lib/puppetserver/ca/puppet_config.rb

```diff
@@ -72,17 +72,33 @@ def resolve_settings(overrides = {})
         ssldir = overrides.fetch(:ssldir, '$confdir/ssl')
         settings[:ssldir] = substitutions['$ssldir'] = ssldir.sub('$confdir', confdir)
 
+        certdir = overrides.fetch(:certdir, '$ssldir/certs')
+        settings[:certdir] = substitutions['$certdir'] = certdir.sub(unresolved_setting, substitutions)
+
         cadir = overrides.fetch(:cadir, '$ssldir/ca')
         settings[:cadir] = substitutions['$cadir'] = cadir.sub(unresolved_setting, substitutions)
 
+        # Note the call to hostname below, this is horrific....
+        settings[:certname] = substitutions['$certname'] = overrides.fetch(:certname, `hostname`.chomp)
```

Comment:
....

Reply:
So.. how we get a certname default seems like it could be problematic.

We definitely shouldn't just use backticks in the middle of a giant method. I'll probably put this behind a method that shells out and move it into the load, parse, or initialize methods (I don't like the idea of doing IO in initialize, but it's also something that should only be done once). So something like:
```rb
        @config_path = supplied_config_path || user_specific_conf_file
        @certname = run("hostname").chomp

```
For this PR `certname` is only need to figure out where `hostcert` lives, and in a default case we could skip certname and figure that `hostcert` is the `.pem` file in `certdir` that _isn't_ `ca.pem`. We could even reverse engineer `certname` from that value... That would avoid shelling out. But I'm unsure of the cases that would or wouldn't work...

Set certname, default hostcert (derived from certname) <- works
Set certname, set hostcert <- works
Set hostcert explicity, default certname <- no good

There's also the dev case of having a bunch of random crap in your certdir from testing, ie mine:
```
11:27:58 puppetserver-ca-cli justin (server2258-revoke-action $) ‖ ls ~/.puppetlabs/etc/puppet/ssl/certs/
bazbar.pem
ca.pem
foobar.pem
localhost.pem
```

Note, in our dev env though we use "localhost" which "hostname" wouldn't guess on my box, we, however, also explictly set certname.

I don't know, I'm thinking shelling out might still be the best bet?

Little rubber ducky-ing, but would love your thoughts, @Magisus @amysahli